[MASWE-0006] Add Sensitive Data Stored Unencrypted in Private Storage Locations (#2566)
OWASP/owasp-mastg: risks/MASVS-STORAGE/1-store-sensitive-data-securely/data-unencrypted-internal/android-data-in-sandbox/test.md

Reviewer: Could you please create an example directory `example-1/` with an example.md and the rest of the required files? (see [here](https://github.com/OWASP/owasp-mastg/tree/master/risks/MASVS-PRIVACY/1-data-minimization/sensitive-data-in-network-traffic/android-data-in-traffic-capture/example-1), and the [guidelines](https://docs.google.com/document/d/1EMsVdfrDBAu0gmjWAUEs60q-fWaOmDB5oecY9d9pOlg/edit?pli=1#heading=h.y294y561hx14))

---
platform: android

> title: Sensitive Data Written to Private Data Directory (Sandbox) Unencrypted

Reviewer: ```suggestion
title: Data Stored to the App Sandbox on Runtime
```

type: [dynamic, filesystem]
mitigations:
- android-use-keystore
- android-use-androidx-security
prerequisites:
- identify-sensitive-data
---

> 

Reviewer: Overview missing.

```suggestion

## Overview

Android apps use a variety of APIs to write data to internal storage. If you only need to inspect the list of created/modified files and their contents, the most effective approach is a snapshot-based approach like the one used in this test.

This has the limitation that you won't know the APIs and locations in your code that are responsible; if you need to know, you should rather follow a dynamic analysis approach based on method tracing. 

```

## Steps

> 1. Start the device.
> 
> 2. Launch and use the app going through the various workflows while inputting sensitive data wherever you can. Taking note of the data you input can help identify it later using tools to search for it.

Reviewer: Aren't we missing this first copy to have as a base for the diff later?

> About the link: I'm aware that this is rather cumbersome, but we are researching a new, simpler solution for this type of linking.

```suggestion
1. Start the device.

2. Take a first [copy of the app's private data directory](../../../../../techniques/android/MASTG-TECH-0008.md) to have as a reference for offline analysis.

3. Launch and use the app going through the various workflows while inputting sensitive data wherever you can. Taking note of the data you input can help identify it later using tools to search for it.
```

> 3. Take a copy of the app's private data directory for offline analysis. Using tar will preserve the filesystem structure and permissions.

Reviewer 2: I wonder if the step 3 should be more detailed, especially for users who are familiar with those type of taks or techniques.

- How would they extract the data? 
- What tool would they used? 
- Where could they find some documentation about this?

Author: My understanding is that the detailed steps and tools would be in the examples under this test, and this just describes the high level process. The examples can then show more than one way or tool. Probably a good questions to clarify in the call next week.

Author: Agreed on the call that the more specific steps would be in the example tests.

> 3. Take a copy of the app's private data directory for offline analysis. See: https://mas.owasp.org/MASTG/techniques/android/MASTG-TECH-0008.

Reviewer: ```suggestion
4. Take a second copy of the app's private data directory for offline analysis and make a diff using the first copy to identify all files created or modify during your testing session.
```

4. Search the extracted data for items such as keys, passwords and any sensitive data inputted into the app.

> 5. Check files for sensitive data that has been encoded with algorithms such as base64 which obscures but does not protect sensitive data.

Reviewer: Maybe step 4 is the actual evaluation, i.e. the final search after everything has been unveiled/decompressed/decoded/deobfuscated etc.

One suggestion for the current step 5 would be:

> Attempt to identify and decode data that has been encoded using methods such as base64 encoding, hexadecimal representation, URL encoding, escape sequences, wide characters and common data obfuscation methods such as xoring. Also consider identifying and decompressing compressed files such as tar or zip. These methods obscure but do not protect sensitive data.

What do you think about this? Maybe we can write a MASTG-TECH for this and keep it short here.

Reviewer: ```suggestion
4. Attempt to identify and decode data that has been encoded using methods such as base64 encoding, hexadecimal representation, URL encoding, escape sequences, wide characters and common data obfuscation methods such as xoring. Also consider identifying and decompressing compressed files such as tar or zip. These methods obscure but do not protect sensitive data.
```

Reviewer: If we apply the above changes this will be step 5.

## Observation

Files within the private data directory contain sensitive data.

## Evaluation

> The test case fails if you find sensitive data in the app's private data directory which has not been encrypted with strong cryptography. This includes plaintext data as well as encoding such as base64 or obfuscation such as xoring.

Reviewer: The [observation](https://docs.google.com/document/d/1EMsVdfrDBAu0gmjWAUEs60q-fWaOmDB5oecY9d9pOlg/edit#heading=h.h9gqgz4hdubj) should describe the resulting data that is going to be evaluated. For example, the "extracted data from the app's private data directory properly decoded, uncompressed, etc."

Then the evaluation could be what step 4 was. Something like:

> Search the extracted data for items such as keys, passwords and any sensitive data inputted into the app. The test case fails if you find any of this sensitive data.
